packages/rpc-transport/src/json-rpc-subscription.ts

```diff
@@ -0,0 +1,142 @@
+                    if ('error' in message) {
+                        throw new SolanaJsonRpcError(message.error);
+                    } else {
+                        subscriptionId = message.result as SubscriptionId;
+                        break;
+                    }
+                }
+            }
+            if (subscriptionId == null) {
+                // TODO: Coded error.
+                throw new Error('Failed to obtain a subscription id from the server');
+            }
+            /**
+             * STEP 3: Return an iterable that yields notifications for this subscription id.
+             */
+            return {
+                async *[Symbol.asyncIterator]() {
+                    for await (const message of connection as AsyncIterable<
+                        JsonRpcNotification<unknown> | JsonRpcResponse<SubscriptionId>
+                    >) {
+                        if (!('params' in message) || message.params.subscription !== subscriptionId) {
+                            continue;
+                        }
+                        const notification = message.params.result as TNotification;
+                        yield responseProcessor ? responseProcessor(notification) : notification;
+                    }
+                },
+            };
+        },
+    };
+}
+
+function makeProxy<TRpcSubscriptionMethods>(
+    rpcConfig: RpcSubscriptionConfig<TRpcSubscriptionMethods>
+): RpcSubscriptions<TRpcSubscriptionMethods> {
+    return new Proxy(rpcConfig.api, {
+        defineProperty() {
+            return false;
+        },
+        deleteProperty() {
+            return false;
+        },
+        get(target, p, receiver) {
+            return function (...rawParams: unknown[]) {
+                const methodName = p.toString();
+                const createRpcSubscription = Reflect.get(target, methodName, receiver);
+                if (p.toString().endsWith('Notifications') === false && !createRpcSubscription) {
+                    // TODO: Coded error.
+                    throw new Error(
+                        "Either the notification name must end in 'Notifications' or the API " +
+                            'must supply a subscription creator function to map between the ' +
+                            'notification name and the subscribe/unsubscribe method names.'
+                    );
+                }
+                const newRequest = createRpcSubscription
+                    ? createRpcSubscription(...rawParams)
+                    : {
+                          params: rawParams,
+                          subscribeMethodName: methodName.replace(/Notifications$/, 'Subscribe'),
+                          unsubscribeMethodName: methodName.replace(/Notifications$/, 'Unsubscribe'),
+                      };
+                return createPendingRpcSubscription(rpcConfig, newRequest);
+            };
+        },
+    }) as RpcSubscriptions<TRpcSubscriptionMethods>;
+}
```

Comment:
Nice, this looks just like the HTTP proxy setup. Do we dare combine them? 😂 😂

Reply:
Sure does. Just with the exception of having one `methodName` for one-shot requests and separate subscribe/unsubscribe method names for subscriptions. Also, the subscriptions one does that janky pattern matching on `*Notifications` thing.

I thought about combining them but thought it would be confusing in the end for little gain.

Reply:
I would have to agree